src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,609 @@
+#else
+        public nint NativeTID;
+        public int ManagedTID;
+        public bool IsMainThread;
+        public JSSynchronizationContext SynchronizationContext;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool IsCurrentThread()
+        {
+            return ManagedTID == Thread.CurrentThread.ManagedThreadId;
+        }
+
+        [UnsafeAccessor(UnsafeAccessorKind.Field, Name = "thread_id")]
+        private static extern ref long GetThreadNativeThreadId(Thread @this);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static IntPtr GetNativeThreadId()
+        {
+            return (int)GetThreadNativeThreadId(Thread.CurrentThread);
+        }
+
+        public JSProxyContext(bool isMainThread, JSSynchronizationContext synchronizationContext)
+        {
+            SynchronizationContext = synchronizationContext;
+            Interop.Runtime.InstallWebWorkerInterop();
+            NativeTID = GetNativeThreadId();
+            ManagedTID = Thread.CurrentThread.ManagedThreadId;
+            IsMainThread = isMainThread;
+        }
+#endif
+
+        #region Current operation context
+
+#if !FEATURE_WASM_THREADS
+        public static readonly JSProxyContext MainThreadContext = new();
+        public static JSProxyContext CurrentThreadContext => MainThreadContext;
+        public static JSProxyContext CurrentOperationContext => MainThreadContext;
+        public static JSProxyContext PushOperationWithCurrentThreadContext()
+        {
+            // in single threaded build we don't have to keep stack of operations and the context/thread is always the same
+            return MainThreadContext;
+        }
+#else
+
+        // Context of the main thread
+        private static JSProxyContext? _MainThreadContext;
+        public static JSProxyContext MainThreadContext
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _MainThreadContext!;
+            set => _MainThreadContext = value;
+        }
+
+        private sealed class PendingOperation
+        {
+            public JSProxyContext? CapturedContext;
+            public bool Called;
+            public bool Multiple;
+        }
+
+        [ThreadStatic]
+        private static List<PendingOperation>? _OperationStack;
+        private static List<PendingOperation> OperationStack
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                return _OperationStack ??= new();
+            }
+        }
+
+        public static void PushOperationUnknownContext()
+        {
+            var stack = OperationStack;
+            // for SchedulePopOperation()
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+            // because this is called multiple times for each JSImport
+            if (stack.Count > 0 && stack[stack.Count - 1].Multiple)
```

Comment:
Should `.Called` and `.Multiple` have different names that more clearly communicate what they mean? The impression I'm getting is that `.Called == true` means 'a pop was scheduled for this stack frame` so I'm not sure why it's named `Called`, and `.Multiple` means 'we expect multiple pushes to happen while this stack frame is active, so ignore any pushes that happen until it is popped'.

Given that I might name them something like `PopScheduled` and `IgnoreAdditionalPushes` respectively.